### ISSUE TITLE
Can Cython work with extern C API which has one parameter used as input and output buffer

### ISSUE THREAD

**Issue by halehawk** [zarr-developers/numcodecs#441]
This is not a bug of numcodecs, I am working on some Cython codes using some extern C APIs. But I met some problems that the C API use one parameter as input and output buffers, do you know if Cython can work with C APIs to send in input array pointer and get the output array from the same pointer? Especially using Buffer and its pointer that is defined in numcodecs @jakirkham

**Comment by halehawk** [comment]
Thank you, I didn't know I could use fromMemory. Currently, I copied the
input numpy array to an output numpy array, and used the output numpy array
Buffer type to call the C API. And it worked. I will try fromMemory later.

**Comment by martindurant** [comment]
Copying is, of course, the safer approach :)

**Comment by halehawk** [comment]
I have another question. If you can help me, that would be great.
I want to add another compressor SPERR in numcodecs, so I followed the way
blosc used, I added SPERR as a submodule and compiled SPERR source codes
and python binding (sperr.pyx) in setup.py.
I can get it to work and I can use it to compile data successfully. But I
cannot get the same results as I use SPERR to compress from its C utility
command. I tried to make sure all C related flags and options were the
same. but I still cannot get the same compress/decompress results
bit-by-bit. So I changed the way to compile, now I compile sperr.pyx only
and link it to SPERR C library. Then I can get the same results bit-by-bit.
Do you know how blosc did it? Or is it supposed to be different?

**Comment by martindurant** [comment]
You mean https://github.com/NCAR/numcodecs_sperr ?

I believe the submodule tactic is complex and there have been conversations about avoiding it. There isn't a real reason for such a dependency to be included directly in numcodecs, rather than in a separate package referenced by an entrypoint.

Some codecs work by compiling the low-level C stuff into a binary and separate python bindings, which works particularly well with conda, which happily installs binary non-python dependencies. The cython part then only needs include statements.

**Comment by halehawk** [comment]
Yes, this is the compressor I intend to add to numcodecs. It can compress
data by specified absolute error value, such as 0.01 for data that is
around that range. The PSNR mode is good too.
Also we have developed a log filter and other filters to use before
compression, I write a Python binding for this log filter, and call it from
SPERR encode, which is good for data that is in a large range. Do you think
I can create a PR for this compressor in numcodecs?

**Comment by martindurant** [comment]
> Do you think I can create a PR for this compressor in numcodecs?

In principle it can be OK, but I think you would find resistance to adding extra compile complexity or submodule into this package directly. Is there a reason not to make it a separate package?

**Comment by halehawk** [comment]
I don't know how to make it a separate package officially. Since SPERR is a
C/C++ package, where can I publish libSPERR.so and then connect it to
sperr.pyx which will be in numcodecs?

**Comment by martindurant** [comment]
Two options:
- sperr.pyx does not need to be in numcodecs, it can live in the same package as the C code
- you can distribute your .so in a python wheel (hard) or a conda package (easy, but not as widely used; common in scientific scenarios, though)

**Comment by martindurant** [comment]
Perhaps check out what imagecodecs does. That includes all the C code in the repo, with bindings to expose both compress/uncompress functions and numcodecs codecs. It shows how you can call register at runtime or establish entrypoints for your codec.
